presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java

```diff
@@ -467,17 +458,35 @@ private static int findGroupEnd(PagesIndex pagesIndex, PagesHashStrategy pagesHa
         checkArgument(pagesIndex.getPositionCount() > 0, "Must have at least one position");
         checkPositionIndex(startPosition, pagesIndex.getPositionCount(), "startPosition out of bounds");
 
-        // Short circuit if the whole page has the same value
-        if (pagesIndex.positionEqualsPosition(pagesHashStrategy, startPosition, pagesIndex.getPositionCount() - 1)) {
-            return pagesIndex.getPositionCount();
+        return findEndPosition(startPosition + 1, pagesIndex.getPositionCount(), (firstPosition, secondPosition) -> pagesIndex.positionEqualsPosition(pagesHashStrategy, firstPosition, secondPosition));
+    }
+
+    // equalityComparator is suppose to return true if positions given as parameters are equal
+    @VisibleForTesting
+    static int findEndPosition(int startingPosition, int lastPosition, BiFunction<Integer, Integer, Boolean> equalityComparator)
+    {
+        if (startingPosition >= lastPosition || !equalityComparator.apply(startingPosition, startingPosition - 1)) {
+            return startingPosition;
         }
 
-        // TODO: do position binary search
-        int endPosition = startPosition + 1;
-        while ((endPosition < pagesIndex.getPositionCount()) &&
-                pagesIndex.positionEqualsPosition(pagesHashStrategy, endPosition - 1, endPosition)) {
-            endPosition++;
+        int diff = lastPosition - startingPosition;
+        int midPosition = startingPosition + floorDiv2(diff);
```

Comment:
I was afraid that Javac may not do this in clever way. According to offline discussion, I'm going to do simple `diff / 2` given Javac will do this clever anyway.